src/expr/sequence.cpp

```diff
@@ -0,0 +1,322 @@
+    if (s == p)
+    {
+      return i;
+    }
+  }
+  return i;
+}
+
+size_t Sequence::roverlap(const Sequence& y) const
+{
+  size_t i = size() < y.size() ? size() : y.size();
+  for (; i > 0; i--)
+  {
+    Sequence s = prefix(i);
+    Sequence p = y.suffix(i);
+    if (s == p)
+    {
+      return i;
+    }
+  }
+  return i;
+}
+
+bool Sequence::isRepeated() const
+{
+  if (size() > 1)
+  {
+    Node f = d_seq[0];
+    for (unsigned i = 1, sz = size(); i < sz; ++i)
+    {
+      if (f != d_seq[i])
+      {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+size_t Sequence::find(const Sequence& y, size_t start) const
+{
+  if (size() < y.size() + start)
+  {
+    return std::string::npos;
+  }
+  if (y.empty())
+  {
+    return start;
+  }
+  if (empty())
+  {
+    return std::string::npos;
+  }
+  std::vector<Node>::const_iterator itr = std::search(
+      d_seq.begin() + start, d_seq.end(), y.d_seq.begin(), y.d_seq.end());
+  if (itr != d_seq.end())
+  {
+    return itr - d_seq.begin();
+  }
+  return std::string::npos;
+}
+
+size_t Sequence::rfind(const Sequence& y, size_t start) const
+{
+  if (size() < y.size() + start)
+  {
+    return std::string::npos;
+  }
+  if (y.empty())
+  {
+    return start;
+  }
+  if (empty())
+  {
+    return std::string::npos;
+  }
+  std::vector<Node>::const_reverse_iterator itr = std::search(
+      d_seq.rbegin() + start, d_seq.rend(), y.d_seq.rbegin(), y.d_seq.rend());
+  if (itr != d_seq.rend())
+  {
+    return itr - d_seq.rbegin();
+  }
+  return std::string::npos;
+}
+
+bool Sequence::hasPrefix(const Sequence& y) const
+{
+  size_t s = size();
+  size_t ys = y.size();
+  if (ys > s)
+  {
+    return false;
+  }
+  for (size_t i = 0; i < ys; i++)
+  {
+    if (d_seq[i] != y.d_seq[i])
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool Sequence::hasSuffix(const Sequence& y) const
+{
+  size_t s = size();
+  size_t ys = y.size();
+  if (ys > s)
+  {
+    return false;
+  }
+  size_t idiff = s - ys;
+  for (size_t i = 0; i < ys; i++)
+  {
+    if (d_seq[i + idiff] != y.d_seq[i])
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+Sequence Sequence::replace(const Sequence& s, const Sequence& t) const
+{
+  Assert(d_type == s.d_type);
+  Assert(d_type == t.d_type);
+  size_t ret = find(s);
+  if (ret != std::string::npos)
+  {
+    std::vector<Node> vec;
+    vec.insert(vec.begin(), d_seq.begin(), d_seq.begin() + ret);
+    vec.insert(vec.end(), t.d_seq.begin(), t.d_seq.end());
+    vec.insert(vec.end(), d_seq.begin() + ret + s.size(), d_seq.end());
+    return Sequence(d_type, vec);
+  }
+  return *this;
+}
+
+Sequence Sequence::substr(size_t i) const
+{
+  Assert(i <= size());
+  std::vector<Node> ret_vec;
+  std::vector<Node>::const_iterator itr = d_seq.begin() + i;
+  ret_vec.insert(ret_vec.end(), itr, d_seq.end());
+  return Sequence(d_type, ret_vec);
+}
+
+Sequence Sequence::substr(size_t i, size_t j) const
+{
+  Assert(i + j <= size());
```

Comment:
```suggestion
  Assert(i >= 0);
  Assert(j >= 0);
  Assert(i + j <= size());
```

Reply:
Actually, those assertions are not really necessary since `size_t` is `unsigned`. My bad.